renormalizer/mps/symbolic_mpo.py

```diff
@@ -0,0 +1,645 @@
+    op1_new_sigma_z = (op1.split_symbol.count("sigma_+") + op1.split_symbol.count("sigma_-")) % 2
+    # similar except by op2
+    op2_new_sigma_z = (op2.split_symbol.count("sigma_+") + op2.split_symbol.count("sigma_-")) % 2
+    # determine the coefficient
+    op1_n_sigma_plus = op1.split_symbol.count("sigma_+")
+    op1_n_sigma_minus = op1.split_symbol.count("sigma_-")
+    assert op1_n_sigma_plus in [0, 1]
+    assert op1_n_sigma_minus in [0, 1]
+    n_permutes = op2_new_sigma_z * (op1_n_sigma_plus + op1_n_sigma_minus)
+    coeff = (-1) ** n_permutes
+    # cancel sigma_z as much as possible
+    def prepend_sigma_z(op: Op):
+        symbol_list = op.split_symbol
+        if symbol_list[0] == "I":
+            assert len(symbol_list) == 1
+            new_op = Op("sigma_z", op.dofs[0], qn=0)
+        elif symbol_list[0] == "sigma_z":
+            if len(symbol_list) == 1:
+                new_op = Op.identity(op.dofs[0])
+            else:
+                new_op = Op(" ".join(symbol_list[1:]), op.dofs[1:], qn=op.qn_list[1:])
+        elif symbol_list[0] == "sigma_+" or symbol_list[0] == "sigma_-":
+            new_op = Op("sigma_z " + op.symbol, [op.dofs[0]] + op.dofs, qn=[0] + op.qn_list)
+        else:
+            assert False
+        return new_op
+    if op1_new_sigma_z:
+        new_op2 = prepend_sigma_z(op2)
+    else:
+        new_op2 = op2
+    if op2_new_sigma_z:
+        new_op1 = prepend_sigma_z(op1)
+    else:
+        new_op1 = op1
+
+    if new_op1 not in op2idx:
+        op2idx[new_op1] = len(primary_ops)
+        primary_ops.append(new_op1)
+    if new_op2 not in op2idx:
+        op2idx[new_op2] = len(primary_ops)
+        primary_ops.append(new_op2)
+
+    return [row[0], op2idx[new_op1], op2idx[new_op2], row[3], row[4]], coeff
+
+
+def table_and_factor_swapped_jw(table, factor, primary_ops: List):
+    # modifies primary_ops in place !!
+    new_table = []
+    new_factor = []
+    op2idx = {op: i for i, op in enumerate(primary_ops)}
+    for row, factor_row in zip(table, factor):
+        new_row, coeff = table_row_swapped_jw(row, primary_ops, op2idx)
+        new_table.append(new_row)
+        new_factor.append(coeff * factor_row)
+    return np.array(new_table), np.array(new_factor)
+
+
+def swap_site(out_ops_list, primary_ops: List, swap_jw: bool):
+    # the MPO at hand is - # - # -
+    # the bond indices are 1, 2, 3 and the site indices are 1 and 2
+    # the operators at each of the bond
+    out_ops1, out_ops2, out_ops3 = out_ops_list
+
+    # the expanded sum-of-product form of the operator represented by the two site MPO
+    # i.e. the expanded form of out_ops3
+    # the number of items is equal to the index of bond 3
+    # each item [factor, out_ops[1]idx, primary_ops @ site 1, primary_ops @ site 2]
+    out_ops3_expanded: List[List[ExpandedOp]] = []
+    for out_op_sum_list in out_ops3:
+        out_ops3_expanded.append(expand_out_op_sum_list(out_ops2, out_op_sum_list))
+        del out_op_sum_list
```

Comment:
Is the `del` necessary?

Reply:
When I wrote the code I find that this statement helps me remember `out_op_sum_list` is a local variable. It indeed seems awkward after a while...

Reply:
ok. I get your point.